presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,440 @@
+                        .collect(toImmutableList());
+                if (results.stream().anyMatch(result -> result.cost.isPresent() && result.cost.get().isUnknown())) {
+                    return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
+                }
+                bestResult = results.stream()
+                        .filter(result -> result.planNode.isPresent())
+                        .min(resultOrdering)
+                        .orElseGet(() -> createCrossJoins(sources, outputSymbols));
+
+                // we need to check for empty cost again in case we created a cross join with unknown cost
+                if (bestResult.cost.get().isUnknown()) {
+                    return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
+                }
+                log.debug("Least cost join was: " + bestResult.planNode.get().toString());
+                memo.put(multiJoinKey, bestResult);
+            }
+            return bestResult;
+        }
+
+        /**
+         * This method produces a cross join of all the sources in multiJoinNode.
+         * This gets called once there are no longer edges between any of the join sources.
+         */
+        private JoinEnumerationResult createCrossJoins(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            return sources.stream()
+                    .map(source -> new JoinEnumerationResult(Optional.of(lookup.getCumulativeCost(source, session, symbolAllocator.getTypes())), Optional.of(source)))
+                    .reduce((result1, result2) -> createCrossJoin(result1, result2, outputSymbols))
+                    .orElseThrow(() -> new IllegalStateException("no cross join created"));
+        }
+
+        private JoinEnumerationResult createCrossJoin(JoinEnumerationResult leftResult, JoinEnumerationResult rightResult, List<Symbol> outputSymbols)
+        {
+            checkArgument(leftResult.cost.isPresent() && rightResult.cost.isPresent(), "costs must be present");
+            if (leftResult.cost.get().isUnknown() || rightResult.cost.get().isUnknown()) {
+                return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
+            }
+            checkArgument(leftResult.planNode.isPresent() && rightResult.planNode.isPresent(), "planNodes must be present");
+            PlanNode leftSource = leftResult.planNode.get();
+            PlanNode rightSource = rightResult.planNode.get();
+            Set<Symbol> leftSymbols = ImmutableSet.copyOf(leftSource.getOutputSymbols());
+            Set<Symbol> rightSymbols = ImmutableSet.copyOf(rightSource.getOutputSymbols());
+            ImmutableList.Builder<Expression> joinPredicatesBuilder = ImmutableList.builder();
+
+            // add join conjucts that were not used for inference
+            StreamSupport.stream(EqualityInference.nonInferrableConjuncts(allFilter).spliterator(), false)
```

Comment:
this section be moved to a separate method. It is duplicated in `createCrossJoin` and `createJoin`